bd-proto-util/src/serialization/types.rs

```diff
@@ -0,0 +1,708 @@
+  V: ProtoFieldDeserialize + Default,
+{
+  type Element = (K, V);
+
+  fn deserialize_element(is: &mut CodedInputStream<'_>) -> Result<Self::Element> {
+    deserialize_map_entry(is)
+  }
+
+  fn add_element(&mut self, (key, value): Self::Element) {
+    self.insert(key, value);
+  }
+}
+
+// Note: Vec<u8> is intentionally NOT implemented here because it has special handling
+// as a bytes field (single blob, not repeated elements).
+// This implementation is for Vec<T> where T != u8, which are true repeated fields.
+impl<T: ProtoType> ProtoType for Vec<T> {
+  fn wire_type() -> WireType {
+    T::wire_type()
+  }
+}
+
+impl<T: ProtoFieldSerialize> ProtoFieldSerialize for Vec<T> {
+  fn compute_size(&self, field_number: u32) -> u64 {
+    self
+      .iter()
+      .map(|item| item.compute_size(field_number))
+      .sum()
+  }
+
+  fn serialize(&self, field_number: u32, os: &mut CodedOutputStream<'_>) -> Result<()> {
+    for item in self {
+      item.serialize(field_number, os)?;
+    }
+    Ok(())
+  }
+}
+
+impl<T> RepeatedFieldDeserialize for Vec<T>
+where
+  T: ProtoFieldDeserialize,
+{
+  type Element = T;
+
+  fn deserialize_element(is: &mut CodedInputStream<'_>) -> Result<Self::Element> {
+    T::deserialize(is)
+  }
+
+  fn add_element(&mut self, element: Self::Element) {
+    self.push(element);
+  }
+}
+
+// Implementation for BTreeSet<T> as a repeated field
+impl<T: ProtoType + Ord> ProtoType for std::collections::BTreeSet<T> {
+  fn wire_type() -> WireType {
+    T::wire_type()
+  }
+}
+
+impl<T: ProtoFieldSerialize + Ord> ProtoFieldSerialize for std::collections::BTreeSet<T> {
+  fn compute_size(&self, field_number: u32) -> u64 {
+    self
+      .iter()
+      .map(|item| item.compute_size(field_number))
+      .sum()
+  }
+
+  fn serialize(&self, field_number: u32, os: &mut CodedOutputStream<'_>) -> Result<()> {
+    for item in self {
+      item.serialize(field_number, os)?;
+    }
+    Ok(())
+  }
+}
+
+impl<T> crate::serialization::RepeatedFieldDeserialize for std::collections::BTreeSet<T>
+where
+  T: ProtoFieldDeserialize + Ord,
+{
+  type Element = T;
+
+  fn deserialize_element(is: &mut CodedInputStream<'_>) -> Result<Self::Element> {
+    T::deserialize(is)
+  }
+
+  fn add_element(&mut self, element: Self::Element) {
+    self.insert(element);
+  }
+}
+
+
+// ============================================================================
+// Special types
+// ============================================================================
+
+// Implementation for LogType enum (int32)
+// `LogType` is from `bd_proto::protos::logging::payload`.
+impl ProtoType for bd_proto::protos::logging::payload::LogType {
+  fn wire_type() -> WireType {
+    WireType::Varint
+  }
+}
+
+impl ProtoFieldSerialize for bd_proto::protos::logging::payload::LogType {
+  fn compute_size(&self, field_number: u32) -> u64 {
+    use protobuf::Enum;
+    let val = self.value();
+    if val == 0 {
+      return 0;
+    }
```

Comment:
I see that you lifted the 0/default check from the original code to here, but it would be nice if we did this generically in this file? Like check if value is == to default value (require default) and then skip size/serializing?